### DISCUSSION TITLE
How to specify METPLUS_OBS_PREPBUFR_MAP

### DISCUSSION THREAD

**Discussion by BinbinZhou-NOAA** [dtcenter/METplus#1417] [Configuration] [answered]
I compared  the file PB2NCConfig_wrapped in METplus4.0.0 and METplus4.1.0-beta5, and found  the explicit definition for obs_prefbufr_map  is removed in beta-5 version, and use METPLUS_OBS_PREPBUFR_MAP  Instead.  

I tried to find how to specify  METPLUS_OBS_PREPBUFR_MAP in the PB2NC conf file but can not find it in the METplus_4.1.0-beta5 online documentation (key =  ,  val = ;   pairs}.  In the PB2NC use case  example of the document,  METPLUS_OBS_PREPBUFR_MAP is not specified. This means there are some default values for all possible variables (key)?   What is the default val for PBL height, ceiling, visibility, total cloud, ect?   I prefer to define the these vals by myself in the conf file.  

Binbin Zhou
EMC

**Comment by georgemccabe** [comment]
[PB2NC_OBS_PREPBUFR_MAP](https://metplus.readthedocs.io/en/develop/Users_Guide/glossary.html#term-PB2NC_OBS_PREPBUFR_MAP) is the METplus variable that is used to set this value. The table at the bottom of this section shows the relationship between the settings: https://metplus.readthedocs.io/en/develop/Users_Guide/wrappers.html#pb2nc-met-conf

![metplus_obs_prepbufr_map](https://user-images.githubusercontent.com/23407799/152839690-72bb04ef-65b6-4b52-97d7-3c325f073839.PNG)

At the top of the PB2NC: MET Configuration section is a link to the default config file where you can see the default value of the obs_prepbufr_map variable:

![obs_prepbufr_map_defaults](https://user-images.githubusercontent.com/23407799/152839912-74bcb240-e3db-42fd-bd5a-adee438bc3d3.PNG)

To add a new key/val pair to this map, you will need to copy the default value and add the new item(s) to the end of the list. You do not need to include the square braces from the value. A multiple line METplus variable value must contain the same indentation at every line of the value. The line directly after must not have any indentation to signify the end of the value.

```
PB2NC_OBS_PREPBUFR_MAP =
   { key = "POB";     val = "PRES";  },
   { key = "QOB";     val = "SPFH";  },
   { key = "TOB";     val = "TMP";   },
   { key = "ZOB";     val = "HGT";   },
   { key = "UOB";     val = "UGRD";  },
   { key = "VOB";     val = "VGRD";  },
   { key = "D_DPT";   val = "DPT";   },
   { key = "D_WDIR";  val = "WDIR";  },
   { key = "D_WIND";  val = "WIND";  },
   { key = "D_RH";    val = "RH";    },
   { key = "D_MIXR";  val = "MIXR";  },
   { key = "D_PRMSL"; val = "PRMSL"; },
   { key = "D_PBL";   val = "PBL";   },
   { key = "D_CAPE";  val = "CAPE";  },
   { key = "TOTALCLOUD";  val = "TCDC"; }

```

It is likely that these new additions to your obs_prepbufr_map will be useful to other users, so it would make sense to add these new key/values to the default config file.

Could you please create a [new MET GitHub issue](https://github.com/dtcenter/MET/issues/new?assignees=&labels=alert%3A+NEED+ACCOUNT+KEY%2C+alert%3A+NEED+MORE+DEFINITION%2C+alert%3A+NEED+PROJECT+ASSIGNMENT%2C+type%3A+enhancement&template=enhancement_request.md&title=) and list the new key/value pairs that are needed to be added to the default?

**Reply by BinbinZhou-NOAA** [reply]
@georgemccabe 
I tested the above setting for PB2NC_OBS_PREPBUFR_MAP, but it seems the val values for new added variables are same as keys.  My settings like as below
PB2NC_OBS_BUFR_VAR_LIST = UOB, VOB, TOB, D_DPT, D_CAPE, D_RH, HOVI , TOCC, CEILING, PMO
PB2NC_OBS_PREPBUFR_MAP =
   { key = "UOB" ; val = "UGRD"; },
   { key = "VOB" ; val = "VGRD"; },
   { key = "TOB" ; val = "TMP"; },
   { key = "QOB" ; val = "SPFH"; },
   { key = "D_DPT" ; val = "DPT"; },
   { key = "D_CAPE"; val = "CAPE"; },
   { key = "D_RH" ; val = "RH"; },
   { key = "HOVI" ; val = "VIS"; },
   { key = "TOCC" ; val = "TCDC"; },
   { key = "CEILING" ; val = "HGT"; },
   { key = "PMO" ; val = "PRMSL"; } 

The new added variables are "HOVI", "TOCC", "CEILING" and "PMO"

The execution is successful.  Using ncdump to look at the obs_var list in the output nc file, show following
 obs_var =
  "TMP",
  "UGRD",
  "VGRD",
  "DPT",
  "RH",
  "CAPE",
  "HOVI",
  "TOCC",
  "CEILING",
  "PMO" ; 

This means the 4 varibales are grabed successfully but the key ~ val mapping for the 4 new added variables is not fine  

I recommend add the last 4 to the default variable list for obs_prepbufr_map dictionary

**Reply by BinbinZhou-NOAA** [reply]
Correction:  The previous run is using METplus-4.0.0. After changing to  METplus-4.1.0-beta5 version, the execution runs into error. The error message says:
Error: new-line character seen in unquoted field - do you need to open the file in universal-newline mode? 
If I remove the entire part: 
PB2NC_OBS_PREPBUFR_MAP =
{ key = "UOB" ; val = "UGRD"; },
{ key = "VOB" ; val = "VGRD"; },
{ key = "TOB" ; val = "TMP"; },
{ key = "QOB" ; val = "SPFH"; },
{ key = "D_DPT" ; val = "DPT"; },
{ key = "D_CAPE"; val = "CAPE"; },
{ key = "D_RH" ; val = "RH"; },
{ key = "HOVI" ; val = "VIS"; },
{ key = "TOCC" ; val = "TCDC"; },
{ key = "CEILING" ; val = "HGT"; },
{ key = "PMO" ; val = "PRMSL"; }

Then the  execution is successful, but no key ~ val match for those new added variables.  So the error may be in how to write the {key~val} in  PB2NC_OBS_PREPBUFR_MAP.  This error  is related to Python opening file mode?

**Reply by georgemccabe** [reply]
Hi Binbin, is each line after the first line indented? Multiple line values must be indented with the same number of spaces to signify that the value should continue. The line after the list should not be indented. If the lines are indented and you are still having issues, you can try putting all of the values on the same line.

**Reply by georgemccabe** [reply]
Binbin, I tried running with the text you posted with each line indented and got strange results. It looks like you have a space between the end quotation mark and the semi-colon for each key.  I removed those spaces and it appeared to work properly:

```
PB2NC_OBS_PREPBUFR_MAP =
  { key = "UOB"; val = "UGRD"; },
  { key = "VOB"; val = "VGRD"; },
  { key = "TOB"; val = "TMP"; },
  { key = "QOB"; val = "SPFH"; },
  { key = "D_DPT"; val = "DPT"; },
  { key = "D_CAPE"; val = "CAPE"; },
  { key = "D_RH"; val = "RH"; },
  { key = "HOVI"; val = "VIS"; },
  { key = "TOCC"; val = "TCDC"; },
  { key = "CEILING"; val = "HGT"; },
  { key = "PMO"; val = "PRMSL"; }
```

**Reply by georgemccabe** [reply]
A space following by a semi-colon is the notation for in-line comments in config files. This is why some of the values were being removed. Adding the semi-colon immediately following the quotation mark treats the text as it should instead of considering it a comment.

**Reply by BinbinZhou-NOAA** [reply]
George,

   I removed the space between the quotation mark and semi-colon, then running is successful. But the key ~val match pairs for those new added 4 variables are not fine, shown as below
 obs_var =
  "TMP",
  "UGRD",
  "VGRD",
  "DPT",
  "RH",
  "CAPE",
  "HOVI",
  "TOCC",
  "CEILING",
  "PMO" ;

The last 4 should be VIS, TCDC, HGT, PRSML

**Reply by malloryprow** [reply]
@BinbinZhou-NOAA I set 

> PB2NC_MET_CONFIG_OVERRIDES = obs_prefbufr_map = [ { key = "POB"; val = "PRES"; }, { key = "QOB"; val = "SPFH"; }, { key = "TOB"; val = "TMP"; }, { key = "UOB"; val = "UGRD"; }, { key = "VOB"; val = "VGRD";  }, { key = "D_DPT"; val = "DPT"; }, { key = "D_WDIR"; val = "WDIR"; }, { key = "D_WIND";  val = "WIND"; }, { key = "D_RH"; val = "RH"; }, { key = "D_MIXR"; val = "MIXR"; }, { key = "D_PBL"; val = "HPBL"; }, { key = "D_PRMSL"; val = "PRMSL"; }, { key = "D_CAPE"; val = "CAPE"; }, { key = "TDO"; val = "DPT"; }, { key = "PMO"; val = "PRMSL"; }, { key = "TOCC"; val = "TCDC"; }, { key = "HOVI"; val = "VIS"; }, { key = "CEILING"; val = "HGT"; }, { key = "MXGS"; val = "GUST";} ];

and in my pb2nc netCDF file I see

>  obs_var =
  "SPFH",
  "TMP",
  "UGRD",
  "VGRD",
  "DPT",
  "RH",
  "PRMSL",
  "PRMSL",
  "GUST",
  "DPT",
  "VIS",
  "HGT",
  "TCDC" ;
}

**Reply by BinbinZhou-NOAA** [reply]
@malloryprow Thanks! I'll test it. What METplus versiosn you used for this case?

**Reply by malloryprow** [reply]
@BinbinZhou-NOAA I am using MET version 10.1.0-beta5 and METplus version 4.1.0-beta5

**Reply by BinbinZhou-NOAA** [reply]
@malloryprow  
  Yes, it works in METplus_10.1.0-beta5! Thanks! 
 I  still hope to update the MET's PB2NC's default config file to inlude PMO (for PRMSL), TDO (for DPT), HOVI (for VIS), CEILING (for HGT), TOCC (for TCDC) to match the field strings used in EMC's forecast files. PMO and TDO are MNEMONIC in PREPBUFR file for mean sea level pressure and dew point. I prefer to use them directly  instead of deriving them for observation data

**Comment by JohnHalleyGotway** [comment, marked as answer]
@BinbinZhou-NOAA, George and I did some more digging and now understand what's going on here. There are a few things to clarify.

1. The reason why it's not working as expected it due to a typo. METplus is defining "obs_prep_bufr_map"... BUT on [this line of the PB2NC config file](https://github.com/dtcenter/MET/blob/221e14813aa77f099536c915f357abc9a325cf42/met/data/config/PB2NCConfig_default#L110), I see that its named "obs_prefbufr_map" with an 'F' instead of a 'P'!

2. Next, we realize that METplus should **NOT** actually be setting obs_prefbufr_map (or obs_prepbufr_map). Based on this description of [obs_bufr_map](https://met.readthedocs.io/en/develop/Users_Guide/config_options.html#obs-bufr-map), it should instead be defining PB2NC_OBS_BUFR_MAP to set obs_bufr_map in MET.

So @georgemccabe will update METplus to support PB2NC_OBS_BUFR_MAP and will removed PB2NC_OBS_PREPBUFR_MAP.

3. With [this issue](https://github.com/dtcenter/MET/issues/2044) for MET version 10.1.0, @hsoh-u will also actually fix that typo, changing "obs_prefbufr_map" to "obs_prepbufr_map". And when he does he'll make sure to keep it backward compatible.

Sorry for all the confusion caused by the typo!

**Reply by BinbinZhou-NOAA** [reply]
@JohnHalleyGotway 

 That's great to find the reason.  If also add following 5 { key ~ val } pairs.  it is more convenient for us to generate our pb2nc conf file.  
   { key = "HOVI";    val = "VIS"; },
   { key = "TOCC";    val = "TCDC"; },
   { key = "CEILING";  val = "HGT"; },
   { key = "PMO";     val = "PRMSL"; } 
   { key = "TDO"; val = "DPT"; }
Dew point and mean sea level pressure can be derived through  D_DPT and D_PRMSL,
But since the NCEP's prepbufr files already have  dew point (TDO)  and  mean sea level pressure (PMO) data, we prefer using the observed  instead of derived data.   

Binbin

**Reply by JohnHalleyGotway** [reply]
@BinbinZhou-NOAA, thanks for providing this list.

However, I'm concerned about this one:
```
{ key = "CEILING"; val = "HGT"; },
```

The model output contains several different types of heights... ceiling, cloud base, cloud top, freezing level. All of these are HGT in the GRIB1/2 outputs but are differentiated by the level "type" value. Renaming "CEILING" obs as "HGT" will likely cause confusion.

I'm also not sure how to handle these 2 you referenced:
```
{ key = "PMO"; val = "PRMSL"; }
{ key = "TDO"; val = "DPT"; }
```

If we add those 2, should we REMOVE these existing ones?
```
   { key = "D_DPT";   val = "DPT";   }
   { key = "D_PRMSL"; val = "PRMSL"; }
```

Or should we just continue writing out PMO and TDO without renaming them and let the user pick which to use based on their config file settings? You can configure METplus to verify fcst "DPT" vs obs "TDO"... or fcst "DPT" vs obs "DPT".

**Reply by BinbinZhou-NOAA** [reply]
@JohnHalleyGotway 
   D_DPT and D_PRMSL can be kept and let user select derived or direct observed.
As for CEILING,  I just consider that the ceiling height in NCEP's model outout grib files is represented by "HGT" and "L0". If this is confused with other heights, using CEILING should be fine: 
{ key = "CEILING";  va l= "CEILING"; } 

Binbin

**Reply by JohnHalleyGotway** [reply]
@BinbinZhou-NOAA, we discussed this topic at the METplus development meeting today. Our primary concern is that we do not want to change the results for users from MET version 10.0.0 to version 10.1.0. Looking at the proposed changes to the PB2NC default configuration file we've discussed, some of them WOULD change results. Below I've listed a summary of what we'd like to change (or not) and the reasoning for it:

1. **Add** HOVI and TOCC to OBS_PREPBUFR_MAP in the default PB2NC config file:
```
{ key = "HOVI"; val = "VIS"; },
{ key = "TOCC"; val = "TCDC"; },
```

Renaming these to VIS and TCDC makes verifying them against GRIB outputs more convenient.

2. **DO NOT** add an entry for CEILING. If the input obs name is what we actually want written to the output, there is no need to rename it. Just leave it as is.

3. **DO NOT** add/modify the settings for D_DPT or D_PRMSL. This is the messy one. If we were to rename PMO to PRMSL and TDO to DPT, that would CHANGE the results for other users. I definitely agree that verifying against the raw obs (PMO or TDO) versus derived obs (D_PRMSL and D_DPT), is a very important consideration. However, right now, in all existing PB2NC output, PRMSL and DPT are actually the derived versions (not the raw ones). We should not repurpose those names in MET version 10.1.0.

However, you still can configure/run PB2NC and Point-Stat to do what you'd like. Just verify forecasts of PRMSL against PMO obs and forecasts of DPT against TDO obs. You'll just need to specify the FCST and OBS variable names in METplus separately. But that usage should work fine.

Or you could, of course, configure METplus to do whatever you'd like. Any settings in the METPLUS_OBS_BUFR_MAP take precedence over the default settings in OBS_PREPBUFR_MAP.

I'd also recommend that the NOAA/EVS team give this some thought. What conventions do you want to define/follow for differentiating between raw observations and derived types? If those conventions do require changes to the PB2NC config file, we could consider making them in a major release, like MET version 11.0.0, rather than a minor one, like version 10.1.0.

**Comment by BinbinZhou-NOAA** [comment]
John,

    Thanks for this update and explanation. . No problem to just add VIS
and TCDC and keep others unchanged.  We can handle this on our side and
won't mess up current MET codes and scripts due to our requests.

Binbin

On Tue, Feb 22, 2022 at 1:03 PM johnhg ***@***.***> wrote:

> @BinbinZhou-NOAA <https://github.com/BinbinZhou-NOAA>, we discussed this
> topic at the METplus development meeting today. Our primary concern is that
> we do not want to change the results for users from MET version 10.0.0 to
> version 10.1.0. Looking at the proposed changes to the PB2NC default
> configuration file we've discussed, some of them WOULD change results.
> Below I've listed a summary of what we'd like to change (or not) and the
> reasoning for it:
>
>    1. *Add* HOVI and TOCC to OBS_PREPBUFR_MAP in the default PB2NC config
>    file:
>
> { key = "HOVI"; val = "VIS"; },
> { key = "TOCC"; val = "TCDC"; },
>
> Renaming these to VIS and TCDC makes verifying them against GRIB outputs
> more convenient.
>
>    1.
>
>    *DO NOT* add an entry for CEILING. If the input obs name is what we
>    actually want written to the output, there is no need to rename it. Just
>    leave it as is.
>    2.
>
>    *DO NOT* add/modify the settings for D_DPT or D_PRMSL. This is the
>    messy one. If we were to rename PMO to PRMSL and TDO to DPT, that would
>    CHANGE the results for other users. I definitely agree that verifying
>    against the raw obs (PMO or TDO) versus derived obs (D_PRMSL and D_DPT), is
>    a very important consideration. However, right now, in all existing PB2NC
>    output, PRMSL and DPT are actually the derived versions (not the raw ones).
>    We should not repurpose those names in MET version 10.1.0.
>
> However, you still can configure/run PB2NC and Point-Stat to do what you'd
> like. Just verify forecasts of PRMSL against PMO obs and forecasts of DPT
> against TDO obs. You'll just need to specify the FCST and OBS variable
> names in METplus separately. But that usage should work fine.
>
> Or you could, of course, configure METplus to do whatever you'd like. Any
> settings in the METPLUS_OBS_BUFR_MAP take precedence over the default
> settings in OBS_PREPBUFR_MAP.
>
> I'd also recommend that the NOAA/EVS team give this some thought. What
> conventions do you want to define/follow for differentiating between raw
> observations and derived types? If those conventions do require changes to
> the PB2NC config file, we could consider making them in a major release,
> like MET version 11.0.0, rather than a minor one, like version 10.1.0.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/1417#discussioncomment-2230684>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQBMPFLYEPLTMX7CH2MH7XDU4PFVZANCNFSM5NX6SACA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Binbin Zhou

IMSG at NOAA/NWS/NCEP/EMC

5830 University Research Ct.

College Park, MD 20740

***@***.***

301-683-3683
